File: sdk/js/src/schema-service.ts
import type { Cratebase } from "./client.js";
import type { CollectionModel } from "./types.js";

/** Manage collection schemas. Requires a superuser token in
 * `client.authStore`. */
export class SchemaService {
  constructor(private readonly client: Cratebase) {}

  async getList(): Promise<CollectionModel[]> {
    return this.client.send<CollectionModel[]>("/api/collections", { method: "GET" });
  }

  async getOne(idOrName: string): Promise<CollectionModel> {
    return this.client.send<CollectionModel>(`/api/collections/${idOrName}`, { method: "GET" });
  }

  async create(data: Omit<CollectionModel, "id" | "created" | "updated">): Promise<CollectionModel> {
    return this.client.send<CollectionModel>("/api/collections", { method: "POST", body: data });
  }

  async update(idOrName: string, data: Omit<CollectionModel, "id" | "created" | "updated">): Promise<CollectionModel> {
    return this.client.send<CollectionModel>(`/api/collections/${idOrName}`, { method: "PATCH", body: data });
  }

  async delete(idOrName: string): Promise<void> {
    await this.client.send<void>(`/api/collections/${idOrName}`, { method: "DELETE" });
  }
}

File: sdk/dart/lib/src/schema_service.dart
import 'client.dart';
import 'types.dart';

/// Manage collection schemas. Requires a superuser token in
/// `client.authStore`.
class SchemaService {
  final CratebaseClient client;

  SchemaService(this.client);

  Future<List<CollectionModel>> getList() async {
    final json = await client.send('/api/collections');
    return (json as List<dynamic>)
        .map((e) => CollectionModel.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<CollectionModel> getOne(String idOrName) async {
    final json = await client.send('/api/collections/$idOrName');
    return CollectionModel.fromJson(json as Map<String, dynamic>);
  }

  Future<CollectionModel> create(CollectionModel data) async {
    final json = await client.send(
      '/api/collections',
      method: 'POST',
      body: data.toCreateJson(),
    );
    return CollectionModel.fromJson(json as Map<String, dynamic>);
  }

  Future<CollectionModel> update(String idOrName, CollectionModel data) async {
    final json = await client.send(
      '/api/collections/$idOrName',
      method: 'PATCH',
      body: data.toCreateJson(),
    );
    return CollectionModel.fromJson(json as Map<String, dynamic>);
  }

  Future<void> delete(String idOrName) async {
    await client.send('/api/collections/$idOrName', method: 'DELETE');
  }
}
